addons/resource/models/resource.py

```diff
@@ -751,17 +790,23 @@ def _get_unavailable_intervals(self, start, end):
         start_datetime = timezone_datetime(start)
         end_datetime = timezone_datetime(end)
         resource_mapping = {}
+        calendar_mapping = defaultdict(lambda: self.env['resource.resource'])
         for resource in self:
-            calendar = resource.calendar_id
-            resource_work_intervals = calendar._work_intervals(start_datetime, end_datetime, resource)
-            resource_work_intervals = [(start, stop) for start, stop, meta in resource_work_intervals]
-            # start + flatten(intervals) + end
-            resource_work_intervals = [start_datetime] + list(chain.from_iterable(resource_work_intervals)) + [end_datetime]
-            # put it back to UTC
-            resource_work_intervals = list(map(lambda dt: dt.astimezone(utc), resource_work_intervals))
-            # pick groups of two
-            resource_work_intervals = list(zip(resource_work_intervals[0::2], resource_work_intervals[1::2]))
-            resource_mapping[resource.id] = resource_work_intervals
+            calendar_mapping[resource.calendar_id] |= resource
+
+        for calendar in calendar_mapping.keys():
+            resources = calendar_mapping[calendar]
+            resources_work_intervals = calendar._work_intervals(start_datetime, end_datetime, resources)
+            for resource in resources:
+                resource_work_intervals = resources_work_intervals[resource.id] if len(resources) > 1 else resources_work_intervals
+                resource_work_intervals = [(start, stop) for start, stop, meta in resource_work_intervals]
+                # start + flatten(intervals) + end
+                resource_work_intervals = [start_datetime] + list(chain.from_iterable(resource_work_intervals)) + [end_datetime]
```

Comment:
```suggestion
                resource_work_intervals = chain([start_datetime], chain.from_iterable(resource_work_intervals), [end_datetime])
```

Reply:
Please don't refactor this code! This does not fix anything.